File: Web-labs/weblab/data/vulnerable_samples/sample.js
// Deliberately vulnerable sample (JavaScript / Node) -- AUTHORIZED lab.
// Each numbered comment marks a PLANTED weakness for the SAST module to detect.
// Do NOT run this; it is a static-analysis scan target only.
const express = require("express");
const { exec } = require("child_process");
const db = require("./db");

const app = express();

// [VULN] hardcoded JWT secret / API key
const JWT_SECRET = "supersecret_jwt_signing_key_do_not_share";
const API_KEY = "AKIAIOSFODNN7EXAMPLE";

app.get("/user", (req, res) => {
  // [VULN] SQL built via template literal -> SQL injection
  db.query(`SELECT * FROM users WHERE id = ${req.query.id}`);
  // [VULN] SQL built via + concatenation -> SQL injection
  db.query("SELECT * FROM logs WHERE u = '" + req.query.u + "'");
});

app.get("/ping", (req, res) => {
  // [VULN] child_process.exec on user input -> command injection
  exec("ping -c 1 " + req.query.host, (e, out) => res.send(out));
});

app.get("/calc", (req, res) => {
  // [VULN] eval / Function on input -> code injection
  const r = eval(req.query.expr);
  const f = new Function("return " + req.query.body);
  res.send(String(r) + f());
});

app.get("/hello", (req, res) => {
  // [VULN] res.send of untrusted input -> reflected XSS
  res.send("<h1>Hello " + req.query.name + "</h1>");
  // [VULN] innerHTML of untrusted input -> XSS (DOM, when bundled to client)
  document.getElementById("out").innerHTML = req.query.html;
});

app.get("/load", (req, res) => {
  // [VULN] require() of user input -> arbitrary module load
  const mod = require(req.query.plugin);
  res.send(mod.name);
});

app.listen(3000);

File: Web-labs/weblab/data/vulnerable_samples/sample.py
# Deliberately vulnerable sample (Python) -- AUTHORIZED security-education lab.
# Each numbered comment marks a PLANTED, well-known weakness for the SAST module
# (weblab/source_analysis.py) to detect. Do NOT run this; it is a scan target.
import os
import pickle
import subprocess
import hashlib

import yaml

API_PASSWORD = "S3cr3t-Hardcoded-Pa55w0rd!"   # [VULN] hardcoded secret/credential


def login(conn, username):
    cursor = conn.cursor()
    # [VULN] SQL built by f-string concatenation -> SQL injection
    cursor.execute(f"SELECT * FROM users WHERE name = '{username}'")
    return cursor.fetchone()


def search(conn, term):
    cursor = conn.cursor()
    # [VULN] SQL built by % / + concatenation -> SQL injection
    cursor.execute("SELECT * FROM docs WHERE body LIKE '%" + term + "%'")
    return cursor.fetchall()


def ping(host):
    # [VULN] os.system / shell=True on user input -> command injection
    os.system("ping -c 1 " + host)
    subprocess.call("nslookup " + host, shell=True)


def load_session(blob):
    # [VULN] pickle.loads on untrusted data -> insecure deserialization / RCE
    return pickle.loads(blob)


def parse_config(text):
    # [VULN] yaml.load without SafeLoader -> arbitrary object construction
    return yaml.load(text)


def calculate(expr):
    # [VULN] eval / exec on input -> code injection
    result = eval(expr)
    exec("audit = %r" % result)
    return result


def store_password(pw):
    # [VULN] md5 used for password hashing -> weak/broken hash
    return hashlib.md5(pw.encode()).hexdigest()
